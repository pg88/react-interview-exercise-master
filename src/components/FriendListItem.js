import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import styles from './FriendListItem.css';
import { FEMALE, MALE } from '../constants/ActionTypes';

class FriendListItem extends Component {

    render() {
        return (
            <li className={ styles.friendListItem }>
                <div className={ styles.friendInfos }>
                    <div>
                        <span>{ this.props.name }</span>
                    </div>
                    <div>
                        <small> { this.props.commonFriends }  friends in common</small>
                        <br/>
                        <i className={classnames('fa', {'fa-male': this.props.gender === MALE,
                        'fa-female': this.props.gender === FEMALE})}></i>
                        <small> { this.props.gender } </small>
                    </div>
                </div>
                <div className={ styles.friendActions }>
                    <button className={`btn btn-default ${ styles.btnAction }`}
                            onClick={() => this.props.starFriend(this.props.id)}>
                        <i className={ classnames('fa', {
                              'fa-star': this.props.starred,
                              'fa-star-o': !this.props.starred
                        })}/>
                    </button>
                    <button className={ `btn btn-default ${ styles.btnAction }` }
                            onClick={() => this.props.deleteFriend(this.props.id)}>
                        <i className="fa fa-trash"/>
                    </button>
                </div>
            </li>
        );
    }

}

FriendListItem.propTypes = {
    starred: PropTypes.bool,
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    gender: PropTypes.string.isRequired,
    starFriend: PropTypes.func.isRequired,
    commonFriends: PropTypes.number.isRequired
};

export default FriendListItem
