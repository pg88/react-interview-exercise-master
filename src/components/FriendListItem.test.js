import React from 'react';
import { shallow, mount } from 'enzyme';
import FriendListItem from './FriendListItem';
import { OTHER, FEMALE, MALE } from '../constants/ActionTypes';

describe('FriendListItem', () => {
    const propsData = {
        id: 1,
        starred: true,
        gender: "Male",
        commonFriends: 123,
        starFriend: jest.fn(),
        deleteFriend: jest.fn(),
        name: 'Theodore Roosevelt'

    };
    it('should render correctly', () => {
        expect(shallow(<FriendListItem {...propsData} />)).toMatchSnapshot();
    });
    it('should have the right props', () => {
        expect(Object.keys(propsData).includes('name','id', 'starred', 'commonFriends', 'gender','starFriend', 'deleteFriend'));
    });
    it('should have a valid "id" on the friend props and be a number', () => {
        expect(propsData.id).toBeDefined();
        expect(propsData.id).toBeGreaterThanOrEqual(1);
    });
    it('should have a valid "name" on the friend props and be not empty', () => {
        expect(propsData.name).toBeDefined();
        expect(propsData.name).toMatch(/^[a-zA-Z\s\.]*$/);
    });
    it('should have a valid "starred" on the friend props and be boolean', () => {
        expect(propsData.starred).toBeDefined();
        expect(propsData.starred).toBeTruthy();
    });
    it('should have a valid "commonFriends" on the friend props and be a number', () => {
        expect(propsData.commonFriends).toBeDefined();
        expect(propsData.commonFriends).toBeGreaterThanOrEqual(1);
    });
    xit('should have a valid "starFriend" on the friend props and be a function', () => {
        const component= <FriendListItem {...propsData}/>;
        const mountedComponent= mount(component);
        FriendListItem.waitUntilRender().then(() => {
            mountedComponent.find('.fa-starred').toBe(1);
            done();

        })
    });
    xit('should have a valid "GENDER" on the friend props and be not empty', () => {
        expect(propsData.gender).toBeDefined();
        expect(propsData.gender).toBeDefined();

    })
});