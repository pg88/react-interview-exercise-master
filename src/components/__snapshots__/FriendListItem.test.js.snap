exports[`FriendListItem should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FriendListItem
    commonFriends={123}
    deleteFriend={[Function]}
    gender="Male"
    id={1}
    name="Theodore Roosevelt"
    starFriend={[Function]}
    starred={true} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className={undefined}>
          <div>
                    <span>
                              Theodore Roosevelt
                    </span>
          </div>
          <div>
                    <small>
                               
                              123
                                friends in common
                    </small>
                    <br />
                    <i
                              className="fa fa-male" />
                    <small>
                               
                              Male
                               
                    </small>
          </div>
</div>,
        <div
          className={undefined}>
          <button
                    className="btn btn-default undefined"
                    onClick={[Function]}>
                    <i
                              className="fa fa-star" />
          </button>
          <button
                    className="btn btn-default undefined"
                    onClick={[Function]}>
                    <i
                              className="fa fa-trash" />
          </button>
</div>,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              <span>
                            Theodore Roosevelt
              </span>
</div>,
            <div>
              <small>
                             
                            123
                              friends in common
              </small>
              <br />
              <i
                            className="fa fa-male" />
              <small>
                             
                            Male
                             
              </small>
</div>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                Theodore Roosevelt
</span>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Theodore Roosevelt",
              },
              "ref": null,
              "rendered": "Theodore Roosevelt",
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <small>
                   
                  123
                    friends in common
</small>,
                <br />,
                <i
                  className="fa fa-male" />,
                <small>
                   
                  Male
                   
</small>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    " ",
                    123,
                    "  friends in common",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  " ",
                  123,
                  "  friends in common",
                ],
                "type": "small",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-male",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    " ",
                    "Male",
                    " ",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  " ",
                  "Male",
                  " ",
                ],
                "type": "small",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn btn-default undefined"
              onClick={[Function]}>
              <i
                            className="fa fa-star" />
</button>,
            <button
              className="btn btn-default undefined"
              onClick={[Function]}>
              <i
                            className="fa fa-trash" />
</button>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star" />,
              "className": "btn btn-default undefined",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-trash" />,
              "className": "btn btn-default undefined",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-trash",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className={undefined}>
            <div>
                        <span>
                                    Theodore Roosevelt
                        </span>
            </div>
            <div>
                        <small>
                                     
                                    123
                                      friends in common
                        </small>
                        <br />
                        <i
                                    className="fa fa-male" />
                        <small>
                                     
                                    Male
                                     
                        </small>
            </div>
</div>,
          <div
            className={undefined}>
            <button
                        className="btn btn-default undefined"
                        onClick={[Function]}>
                        <i
                                    className="fa fa-star" />
            </button>
            <button
                        className="btn btn-default undefined"
                        onClick={[Function]}>
                        <i
                                    className="fa fa-trash" />
            </button>
</div>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                <span>
                                Theodore Roosevelt
                </span>
</div>,
              <div>
                <small>
                                 
                                123
                                  friends in common
                </small>
                <br />
                <i
                                className="fa fa-male" />
                <small>
                                 
                                Male
                                 
                </small>
</div>,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  Theodore Roosevelt
</span>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Theodore Roosevelt",
                },
                "ref": null,
                "rendered": "Theodore Roosevelt",
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <small>
                     
                    123
                      friends in common
</small>,
                  <br />,
                  <i
                    className="fa fa-male" />,
                  <small>
                     
                    Male
                     
</small>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      " ",
                      123,
                      "  friends in common",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    " ",
                    123,
                    "  friends in common",
                  ],
                  "type": "small",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "fa fa-male",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      " ",
                      "Male",
                      " ",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    " ",
                    "Male",
                    " ",
                  ],
                  "type": "small",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn btn-default undefined"
                onClick={[Function]}>
                <i
                                className="fa fa-star" />
</button>,
              <button
                className="btn btn-default undefined"
                onClick={[Function]}>
                <i
                                className="fa fa-trash" />
</button>,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star" />,
                "className": "btn btn-default undefined",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-trash" />,
                "className": "btn btn-default undefined",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-trash",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
